// @ng-dynamic-forms/ui-basic/esm5/src/checkbox/dynamic-basic-checkbox.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicCheckboxModel, DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, } from "@ng-dynamic-forms/core";
var DynamicBasicCheckboxComponent = /** @class */ (function (_super) {
    __extends(DynamicBasicCheckboxComponent, _super);
    function DynamicBasicCheckboxComponent(layoutService, validationService) {
        var _this = _super.call(this, layoutService, validationService) || this;
        _this.layoutService = layoutService;
        _this.validationService = validationService;
        _this.blur = new EventEmitter();
        _this.change = new EventEmitter();
        _this.focus = new EventEmitter();
        return _this;
    }
    DynamicBasicCheckboxComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-basic-checkbox",
                    template: "<label [formGroup]=\"group\" [ngClass]=\"[getClass('element', 'control'), getClass('grid', 'control')]\">\n\n    <input type=\"checkbox\"\n           [checked]=\"model.checked\"\n           [formControlName]=\"model.id\"\n           [id]=\"elementId\"\n           [indeterminate]=\"model.indeterminate\"\n           [name]=\"model.name\"\n           [required]=\"model.required\"\n           [tabindex]=\"model.tabIndex\"\n           [value]=\"model.value\"\n           (blur)=\"onBlur($event)\"\n           (change)=\"onChange($event)\"\n           (focus)=\"onFocus($event)\"/><span [ngClass]=\"getClass('element', 'label')\"\n                                            [innerHTML]=\"model.label\"></span></label>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    DynamicBasicCheckboxComponent.ctorParameters = function () { return [
        { type: DynamicFormLayoutService },
        { type: DynamicFormValidationService }
    ]; };
    DynamicBasicCheckboxComponent.propDecorators = {
        group: [{ type: Input }],
        layout: [{ type: Input }],
        model: [{ type: Input }],
        blur: [{ type: Output }],
        change: [{ type: Output }],
        focus: [{ type: Output }]
    };
    return DynamicBasicCheckboxComponent;
}(DynamicFormControlComponent));
export { DynamicBasicCheckboxComponent };
if (false) {
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.group;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.layout;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.model;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.blur;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.change;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.focus;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.layoutService;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.validationService;
}
