// @ng-dynamic-forms/ui-basic/esm2015/src/select/dynamic-basic-select.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, DynamicSelectModel, } from "@ng-dynamic-forms/core";
export class DynamicBasicSelectComponent extends DynamicFormControlComponent {
    /**
     * @param {?} layoutService
     * @param {?} validationService
     */
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
        this.blur = new EventEmitter();
        this.change = new EventEmitter();
        this.focus = new EventEmitter();
    }
}
DynamicBasicSelectComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-basic-select",
                template: "<ng-container [formGroup]=\"group\">\n\n    <select [compareWith]=\"model.compareWithFn\"\n            [formControlName]=\"model.id\"\n            [id]=\"elementId\"\n            [name]=\"model.name\"\n            [ngClass]=\"[getClass('element', 'control'), getClass('grid', 'control')]\"\n            [required]=\"model.required\"\n            [tabindex]=\"model.tabIndex\"\n            (blur)=\"onBlur($event)\"\n            (change)=\"onChange($event)\"\n            (focus)=\"onFocus($event)\">\n\n        <option *ngFor=\"let option of model.options$ | async\"\n                [disabled]=\"option.disabled\"\n                [ngValue]=\"option.value\">{{ option.label }}\n        </option>\n\n    </select>\n\n</ng-container>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
DynamicBasicSelectComponent.ctorParameters = () => [
    { type: DynamicFormLayoutService },
    { type: DynamicFormValidationService }
];
DynamicBasicSelectComponent.propDecorators = {
    group: [{ type: Input }],
    layout: [{ type: Input }],
    model: [{ type: Input }],
    blur: [{ type: Output }],
    change: [{ type: Output }],
    focus: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DynamicBasicSelectComponent.prototype.group;
    /** @type {?} */
    DynamicBasicSelectComponent.prototype.layout;
    /** @type {?} */
    DynamicBasicSelectComponent.prototype.model;
    /** @type {?} */
    DynamicBasicSelectComponent.prototype.blur;
    /** @type {?} */
    DynamicBasicSelectComponent.prototype.change;
    /** @type {?} */
    DynamicBasicSelectComponent.prototype.focus;
    /** @type {?} */
    DynamicBasicSelectComponent.prototype.layoutService;
    /** @type {?} */
    DynamicBasicSelectComponent.prototype.validationService;
}
