// @ng-dynamic-forms/ui-basic/esm2015/src/checkbox/dynamic-basic-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicCheckboxModel, DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, } from "@ng-dynamic-forms/core";
export class DynamicBasicCheckboxComponent extends DynamicFormControlComponent {
    /**
     * @param {?} layoutService
     * @param {?} validationService
     */
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
        this.blur = new EventEmitter();
        this.change = new EventEmitter();
        this.focus = new EventEmitter();
    }
}
DynamicBasicCheckboxComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-basic-checkbox",
                template: "<label [formGroup]=\"group\" [ngClass]=\"[getClass('element', 'control'), getClass('grid', 'control')]\">\n\n    <input type=\"checkbox\"\n           [checked]=\"model.checked\"\n           [formControlName]=\"model.id\"\n           [id]=\"elementId\"\n           [indeterminate]=\"model.indeterminate\"\n           [name]=\"model.name\"\n           [required]=\"model.required\"\n           [tabindex]=\"model.tabIndex\"\n           [value]=\"model.value\"\n           (blur)=\"onBlur($event)\"\n           (change)=\"onChange($event)\"\n           (focus)=\"onFocus($event)\"/><span [ngClass]=\"getClass('element', 'label')\"\n                                            [innerHTML]=\"model.label\"></span></label>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
DynamicBasicCheckboxComponent.ctorParameters = () => [
    { type: DynamicFormLayoutService },
    { type: DynamicFormValidationService }
];
DynamicBasicCheckboxComponent.propDecorators = {
    group: [{ type: Input }],
    layout: [{ type: Input }],
    model: [{ type: Input }],
    blur: [{ type: Output }],
    change: [{ type: Output }],
    focus: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.group;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.layout;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.model;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.blur;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.change;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.focus;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.layoutService;
    /** @type {?} */
    DynamicBasicCheckboxComponent.prototype.validationService;
}
