// @ng-dynamic-forms/ui-basic/esm2015/src/textarea/dynamic-basic-textarea.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, DynamicTextAreaModel } from "@ng-dynamic-forms/core";
export class DynamicBasicTextAreaComponent extends DynamicFormControlComponent {
    /**
     * @param {?} layoutService
     * @param {?} validationService
     */
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
        this.blur = new EventEmitter();
        this.change = new EventEmitter();
        this.focus = new EventEmitter();
    }
}
DynamicBasicTextAreaComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-basic-textarea",
                template: "<ng-container [formGroup]=\"group\">\n\n    <textarea [cols]=\"model.cols\"\n              [formControlName]=\"model.id\"\n              [id]=\"elementId\"\n              [maxlength]=\"model.maxLength\"\n              [minlength]=\"model.minLength\"\n              [name]=\"model.name\"\n              [ngClass]=\"[getClass('element', 'control'), getClass('grid', 'control')]\"\n              [placeholder]=\"model.placeholder\"\n              [readonly]=\"model.readOnly\"\n              [required]=\"model.required\"\n              [rows]=\"model.rows\"\n              [spellcheck]=\"model.spellCheck\"\n              [tabindex]=\"model.tabIndex\"\n              [wrap]=\"model.wrap\"\n              (blur)=\"onBlur($event)\"\n              (change)=\"onChange($event)\"\n              (focus)=\"onFocus($event)\"></textarea>\n\n</ng-container>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
DynamicBasicTextAreaComponent.ctorParameters = () => [
    { type: DynamicFormLayoutService },
    { type: DynamicFormValidationService }
];
DynamicBasicTextAreaComponent.propDecorators = {
    group: [{ type: Input }],
    layout: [{ type: Input }],
    model: [{ type: Input }],
    blur: [{ type: Output }],
    change: [{ type: Output }],
    focus: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DynamicBasicTextAreaComponent.prototype.group;
    /** @type {?} */
    DynamicBasicTextAreaComponent.prototype.layout;
    /** @type {?} */
    DynamicBasicTextAreaComponent.prototype.model;
    /** @type {?} */
    DynamicBasicTextAreaComponent.prototype.blur;
    /** @type {?} */
    DynamicBasicTextAreaComponent.prototype.change;
    /** @type {?} */
    DynamicBasicTextAreaComponent.prototype.focus;
    /** @type {?} */
    DynamicBasicTextAreaComponent.prototype.layoutService;
    /** @type {?} */
    DynamicBasicTextAreaComponent.prototype.validationService;
}
