// @ng-dynamic-forms/ui-basic/esm5/src/radio-group/dynamic-basic-radio-group.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, DynamicRadioGroupModel } from "@ng-dynamic-forms/core";
var DynamicBasicRadioGroupComponent = /** @class */ (function (_super) {
    __extends(DynamicBasicRadioGroupComponent, _super);
    function DynamicBasicRadioGroupComponent(layoutService, validationService) {
        var _this = _super.call(this, layoutService, validationService) || this;
        _this.layoutService = layoutService;
        _this.validationService = validationService;
        _this.blur = new EventEmitter();
        _this.change = new EventEmitter();
        _this.focus = new EventEmitter();
        return _this;
    }
    DynamicBasicRadioGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-basic-radio-group",
                    template: "<fieldset role=\"radiogroup\"\n          [formGroup]=\"group\"\n          [id]=\"elementId\"\n          [name]=\"model.name\"\n          [ngClass]=\"[getClass('element', 'control'), getClass('grid', 'control')]\"\n          [tabindex]=\"model.tabIndex\"\n          (change)=\"onChange($event)\">\n\n    <legend *ngIf=\"model.legend\" [innerHTML]=\"model.legend\"></legend>\n\n    <label *ngFor=\"let option of model.options$ | async\" [ngClass]=\"getClass('element', 'option')\">\n\n        <input type=\"radio\"\n               [formControlName]=\"model.id\"\n               [name]=\"model.name\"\n               [value]=\"option.value\"\n               (blur)=\"onBlur($event)\"\n               (focus)=\"onFocus($event)\"/><span [innerHTML]=\"option.label\"></span></label>\n</fieldset>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    DynamicBasicRadioGroupComponent.ctorParameters = function () { return [
        { type: DynamicFormLayoutService },
        { type: DynamicFormValidationService }
    ]; };
    DynamicBasicRadioGroupComponent.propDecorators = {
        group: [{ type: Input }],
        layout: [{ type: Input }],
        model: [{ type: Input }],
        blur: [{ type: Output }],
        change: [{ type: Output }],
        focus: [{ type: Output }]
    };
    return DynamicBasicRadioGroupComponent;
}(DynamicFormControlComponent));
export { DynamicBasicRadioGroupComponent };
if (false) {
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.group;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.layout;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.model;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.blur;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.change;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.focus;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.layoutService;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.validationService;
}
