// @ng-dynamic-forms/ui-basic/esm2015/src/radio-group/dynamic-basic-radio-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, DynamicRadioGroupModel } from "@ng-dynamic-forms/core";
export class DynamicBasicRadioGroupComponent extends DynamicFormControlComponent {
    /**
     * @param {?} layoutService
     * @param {?} validationService
     */
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
        this.blur = new EventEmitter();
        this.change = new EventEmitter();
        this.focus = new EventEmitter();
    }
}
DynamicBasicRadioGroupComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-basic-radio-group",
                template: "<fieldset role=\"radiogroup\"\n          [formGroup]=\"group\"\n          [id]=\"elementId\"\n          [name]=\"model.name\"\n          [ngClass]=\"[getClass('element', 'control'), getClass('grid', 'control')]\"\n          [tabindex]=\"model.tabIndex\"\n          (change)=\"onChange($event)\">\n\n    <legend *ngIf=\"model.legend\" [innerHTML]=\"model.legend\"></legend>\n\n    <label *ngFor=\"let option of model.options$ | async\" [ngClass]=\"getClass('element', 'option')\">\n\n        <input type=\"radio\"\n               [formControlName]=\"model.id\"\n               [name]=\"model.name\"\n               [value]=\"option.value\"\n               (blur)=\"onBlur($event)\"\n               (focus)=\"onFocus($event)\"/><span [innerHTML]=\"option.label\"></span></label>\n</fieldset>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
DynamicBasicRadioGroupComponent.ctorParameters = () => [
    { type: DynamicFormLayoutService },
    { type: DynamicFormValidationService }
];
DynamicBasicRadioGroupComponent.propDecorators = {
    group: [{ type: Input }],
    layout: [{ type: Input }],
    model: [{ type: Input }],
    blur: [{ type: Output }],
    change: [{ type: Output }],
    focus: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.group;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.layout;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.model;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.blur;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.change;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.focus;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.layoutService;
    /** @type {?} */
    DynamicBasicRadioGroupComponent.prototype.validationService;
}
