// @ng-dynamic-forms/ui-basic/esm2015/src/form-array/dynamic-basic-form-array.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormArrayComponent, DynamicFormArrayModel, DynamicFormLayoutService, DynamicFormValidationService } from "@ng-dynamic-forms/core";
export class DynamicBasicFormArrayComponent extends DynamicFormArrayComponent {
    /**
     * @param {?} layoutService
     * @param {?} validationService
     */
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
        this.blur = new EventEmitter();
        this.change = new EventEmitter();
        this.customEvent = new EventEmitter();
        this.focus = new EventEmitter();
    }
}
DynamicBasicFormArrayComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-basic-form-array",
                template: "<ng-container [formGroup]=\"group\">\n\n    <div [formArrayName]=\"model.id\" [id]=\"elementId\" [ngClass]=\"getClass('element', 'control')\">\n\n        <div *ngFor=\"let groupModel of model.groups; let idx = index\" role=\"group\"\n             [formGroupName]=\"idx\" [ngClass]=\"[getClass('element', 'group'), getClass('grid', 'group')]\">\n\n            <ng-container *ngTemplateOutlet=\"startTemplate?.templateRef; context: groupModel\"></ng-container>\n\n            <dynamic-basic-form-control *ngFor=\"let _model of groupModel.group\"\n                                        [context]=\"groupModel\"\n                                        [group]=\"control.at(idx)\"\n                                        [hidden]=\"_model.hidden\"\n                                        [layout]=\"layout\"\n                                        [model]=\"_model\"\n                                        [templates]=\"templates\"\n                                        [ngClass]=\"[getClass('element', 'host', _model), getClass('grid', 'host', _model),\n                                                    getClass('element','children'), getClass('grid','children')]\"\n                                        (blur)=\"onBlur($event)\"\n                                        (change)=\"onChange($event)\"\n                                        (focus)=\"onFocus($event)\"\n                                        (bsEvent)=\"onCustomEvent($event, null, true)\"></dynamic-basic-form-control>\n\n            <ng-container *ngTemplateOutlet=\"endTemplate?.templateRef; context: groupModel\"></ng-container>\n\n        </div>\n\n    </div>\n\n</ng-container>\n"
            }] }
];
/** @nocollapse */
DynamicBasicFormArrayComponent.ctorParameters = () => [
    { type: DynamicFormLayoutService },
    { type: DynamicFormValidationService }
];
DynamicBasicFormArrayComponent.propDecorators = {
    group: [{ type: Input }],
    layout: [{ type: Input }],
    model: [{ type: Input }],
    templates: [{ type: Input }],
    blur: [{ type: Output }],
    change: [{ type: Output }],
    customEvent: [{ type: Output }],
    focus: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DynamicBasicFormArrayComponent.prototype.group;
    /** @type {?} */
    DynamicBasicFormArrayComponent.prototype.layout;
    /** @type {?} */
    DynamicBasicFormArrayComponent.prototype.model;
    /** @type {?} */
    DynamicBasicFormArrayComponent.prototype.templates;
    /** @type {?} */
    DynamicBasicFormArrayComponent.prototype.blur;
    /** @type {?} */
    DynamicBasicFormArrayComponent.prototype.change;
    /** @type {?} */
    DynamicBasicFormArrayComponent.prototype.customEvent;
    /** @type {?} */
    DynamicBasicFormArrayComponent.prototype.focus;
    /** @type {?} */
    DynamicBasicFormArrayComponent.prototype.layoutService;
    /** @type {?} */
    DynamicBasicFormArrayComponent.prototype.validationService;
}
